grammar.js

```diff
@@ -3124,8 +3126,18 @@ module.exports = grammar({
       ),
     ),
     _natural_number: _ => /\d+/,
-    _integer: $ => seq(optional("-"), $._natural_number),
-    _decimal_number: $ => seq(optional("-"), /(\d*[.]\d+)|(\d+[.])/),
+    _integer: $ => seq(
+      optional(choice("-", "+")),
+      /(0[xX][0-9A-Fa-f]+(_[0-9A-Fa-f]+)*)|(0[oO][0-7]+(_[0-7]+)*)|(0[bB][01]+(_[01]+)*)|(\d+(_\d+)*(e[+-]?\d+(_\d+)*)?)/
+    ),
+    _decimal_number: $ => seq(
+      optional(
+        choice("-", "+")),
+      /((\d+(_\d+)*)?[.]\d+(_\d+)*(e[+-]?\d+(_\d+)*)?)|(\d+(_\d+)*[.](e[+-]?\d+(_\d+)*)?)/
+    ),
+    _bit_string: $ => seq(/[bBxX]'([^']|'')*'/, repeat(/'([^']|'')*'/)),
+    // The identifier should be followed by a string (no parenthesis allowed)
+    _string_casting: $ => seq($.identifier, $._single_quote_string),
```

Comment:
This is essentially the same as `implicit_cast`. Could that be added there as an alternative?

Reply:
In spirit it is a literal and not a cast, it only work with a type followed by a literal string.
It is not a cast because I think it is done at "compile time".
```
postgres=# select REAL '1.25';
 float4 
--------
   1.25
(1 row)

postgres=# select REAL '' ;
ERROR:  invalid input syntax for type real: ""
LINE 1: select REAL '' ;
                    ^
postgres=# select REAL '' || '1.25';
ERROR:  invalid input syntax for type real: ""
LINE 1: select REAL '' || '1.25';
                    ^
postgres=# select REAL ('' || '1.25');
ERROR:  syntax error at or near "("
LINE 1: select REAL ('' || '1.25');
                    ^
postgres=# select REAL ('1.25');
ERROR:  syntax error at or near "("
LINE 1: select REAL ('1.25');
                    ^
postgres=# select REAL '1.25';
 float4 
--------
   1.25
(1 row)

postgres=# select REAL column1 from (values('1.25'));
ERROR:  column "real" does not exist
LINE 1: select REAL column1 from (values('1.25'));
               ^
```